[apollo-tan] iP

apollo-tan opened the pull request:
# Kuki Shinobu 🍪🥷 
> “Not only should you have a to-do list, but it must become your best friend.” — Jim Kwik

Often find yourselves feeling overwhelmed and missing deadlines, yet not sure how to organise your life? Look no further, your best friend from Genshin Impact, Kuki Shinobu, is here to hold you accountable. 

Why Kuki Shinobu? Because unlike other todo list apps, Kuki Shinobu offers:
- A **minimalistic UI** that allows you to focus on the tasks that matters 👍 
- **Intuitive commands** that makes it very easy to learn how to use the application 💻 
- An element of **gamification**, making task management more enjoyable 🎮 
## Setting Up
1. Navigate to the [GitHub Repository](https://github.com/apollo-tan/ip)
2. [Clone](https://git-scm.com/docs/git-clone) the repository onto your local machine
3. Run the `main` method of `KukiShinobu` class
4. Type the commands in the terminal and let the application reintroduce order to your life
## Features
- [x] Adding, finding and deleting tasks
- [x] Adding tasks with deadlines or start/end date
- [ ] Modifying Existing tasks (coming soon)
- [ ] Enhanced UI (coming soon)
- [ ] Ability to speak to other Genshin Impact Characters (coming soon)

## Advanced
If you're tech savvy enough, you can also contribute to the code! Here's a snippet of the `main` method of `KukiShinobu` class:
```java
public static void main(String[] args) {  
    KukiShinobu kukiShinobu = new KukiShinobu(FILE_PATH);  
    kukiShinobu.listen();  
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

apollo-tan committed:
Add greet and exit feature

apollo-tan committed:
Add echo feature to chatbot

- add listen instance method to accept user input from commandline
- refactor main to call greet and goodbye in listen method

apollo-tan committed:
Add support for add and list tasks

- Refactor "listen" instance method to use switch-case to determine command
- Add instance field ArrayList "tasks" to store tasks
- Add instance method "listTasks" to list all tasks
- Add instance method "addTask" to add tasks to private instance field "tasks"

apollo-tan committed:
Add support for mark as done

- Create new class Task to abstract the behaviour and properties of task
- Add ability to split user input into command and arguments to support mark as done feature
- Add extra case statements to switch block in "listen" to handle "mark" and "unmark" commands
- Rename kuki to shinobu for clarity
- Convert instance field "tasks" from ArrayList<String> to ArrayList<Task> to support new Task class
- Modify listTasks to account for change in string representation of task

apollo-tan committed:
Add support for different task types

- Add Todo, Deadline and Event subclasses to support different task types
- Convert Task to an abstract class
- Shift arguments handling into the command handlers themselves
- Add addTodo, addDeadline and addEvent instance methods to handle new
  commands

apollo-tan committed:
Configure automated Text UI testing

apollo-tan committed:
Add support for error handling

- Add custom checked exception KukiShinobuException for error handling
- Modify "listen"  to handle checked exception KukiShinobuException
- Add error handling for misconfigured arguments for the various commands

apollo-tan committed:
Add support for delete

- Add deleteTask and printDeletedTaskSummary methods
- Modify logic of markAsDone and unmarkAsDone to be similar to deleteTask
- Modify greeting to better reflect true personality of Kuki Shinobu

apollo-tan committed:
Add support for file reading feature

- Overload constructor for `Task`, `Todo`, `Deadline` and `Event` classes with an additional method signature that accepts an additional `isDone` argument to allow direct instantiation of the aforementioned classes with its `isDone` field set to `true`.
- Add `loadDatabaseFile()`, `parseTasks` and `readFile` methods to `KukiShinobu` to handle file reading, parsing and error handling.
- Modify `listen` public method in `KukiShinobu` to update the file contents for every iteration of the while loop to ensure that it is kept up to date.

apollo-tan committed:
Add support for file writing feature

- Add "getDatabaseString" methods in Task and its subclasses to facilitate writing to database text file
- Add `writeToDatabaseFile` method to write updated tasks to file every time the user performs a command

apollo-tan committed:
Merge branch 'branch-Level-7'

apollo-tan committed:
Add support for date feature

- Add public static methods `getDateStringPrintFormat` and `getDateStringStorageFormat` to parent abstract class `Task` to abstract the various string representations of `LocalDate` required in our application
- Modify date representation in `Deadline` and `Event` classes

apollo-tan committed:
Merge branch 'branch-Level-8'

apollo-tan committed:
Refactor code to support MoreOOP and Packages

- Create command, exception, parser, storage, task and ui packages so that Command and its subclasses can work
- Add Command abstract class and subclasses for each command to abstract away the logic
- Add Parse, Storage, TaskList and Ui classes to abstract the logic for parsing command, file read/write, managing tasks and user input/output respectively
- Revamp KukiShinobu main class to use the new classes and removed all unnecessary code

apollo-tan committed:
Merge branch 'A-MoreOOP'

apollo-tan committed:
Merge branch 'add-gradle-support'

apollo-tan committed:
Add tests for Parser and TaskList classes

apollo-tan committed:
Merge branch 'A-JUnit'

apollo-tan committed:
Modify build.gradle to facilitate building of jar file

apollo-tan committed:
Merge branch 'A-Jar'

limyuquan commented on src/main/java/main/KukiShinobu.java:
```diff
@@ -0,0 +1,63 @@
+import storage.Storage;
+import task.*;
+import ui.Ui;
+import command.*;
```
write full method names rather than using wildcard

limyuquan commented on src/main/java/main/KukiShinobu.java:
```diff
@@ -0,0 +1,63 @@
+import exception.KukiShinobuException;
+import parser.Parser;
+import storage.Storage;
+import task.*;
```
write full method names rather than using wildcard

limyuquan commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,54 @@
+package parser;
+
+import command.*;
```
write full method names rather than using wildcard

limyuquan commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,54 @@
+        //prevents index out of bounds error if no arguments provided after command
+        String arguments = parts.length > 1 ? parts[1] : "";
+
+        // break out of while loop if user issues "bye" command
```
i think this comment's associated code has already been deleted

limyuquan commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,128 @@
+
+    // TODO: load method that reads from the specified file path and returns a task list
+    public ArrayList<Task> load() throws FileNotFoundKukiShinobuException {
+        // Step 1: Read the file, and throw an error is the file can't be found
```
I like the comments to anotate

limyuquan commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,128 @@
+            System.err.println("Error creating or accessing the file: " + filePath);
+            e.printStackTrace();
+            throw new FileNotFoundKukiShinobuException();
+//            return new ArrayList<task.Task>();
```
Could be cleaned up

limyuquan commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,128 @@
+            throw new FileNotFoundKukiShinobuException();
+//            return new ArrayList<task.Task>();
+        }
+//        System.out.println(input);
```
Could be cleaned up

limyuquan commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,128 @@
+        }
+
+        // Step 2: Insert the entire text string into the file
+        FileWriter fw = null;
```
fw variable not directly clear

limyuquan reviewed (commented):
I think some of the comments could be cleaned up, but other than that lgtm

wubojin commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    public static String getDateStringPrintFormat(LocalDate date) {
+        return date.format(DateTimeFormatter.ofPattern("MMM d yyyy"));
```
May consider defining the date formats as constant in Task and use it in it's subclasses like Deadline, Event.
For example,
``` java
protected static final String DATE_PRINT_FORMAT = "MMM d yyyy";
```

wubojin commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,50 @@
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
+public abstract class Task {
```
It would be helpful to add JavaDoc comments to your non-trivial classes and methods.

wubojin commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,50 @@
+        return (isDone ? "X" : " "); // mark done task with X
+    }
+
+    public void markAsDone() {
```
I like that your method names (`markAsDone`, `unmarkAsDone`, `getDatabaseString`) are well chosen and follow the camelCase convention 👍

wubojin commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,50 @@
+
+public abstract class Task {
+    protected String description;
+    protected boolean isDone;
```
It is nice to see that fields in `Task`, `Todo`, `Deadline`, `Event` are `protected` instead of simply `private`.

wubojin commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,59 @@
+package task;
+
+import java.util.ArrayList;
```
Try to include spacing after `import` statements and before the class declaration to improve code readability.

wubojin commented on src/main/java/ui/Ui.java:
```diff
@@ -0,0 +1,55 @@
+    private Scanner scanner = new Scanner(System.in);
+
+
+    // Reads the user-input command and returns it to the program
```
JavaDoc comments should have the following [form](https://nus-cs2103-ay2425s1.github.io/website/se-book-adapted/chapters/documentation.html#javadoc)

wubojin reviewed (commented):
Generally follows coding standards, such as naming and layout. But need to take note of import statements and comments.

apollo-tan committed:
Modify code more closely adhere to the style guide after comments from peers

- remove unnecessary commented out code and annotations
- replace all wildcard imports with explicit imports
- rename `fw` to `filewriter` for clarity

apollo-tan committed:
Add header comments to Command and all its subclasses

apollo-tan committed:
Add header comments to all files in the exception package

apollo-tan committed:
Add header comments to KukiShinobu, Parser and Storage classes

apollo-tan committed:
Add header comments to all files in task and ui packages

apollo-tan committed:
Fix styling issues in several parts of the code

apollo-tan committed:
Add support for find

- Add `findMatchingTasks` method to `TaskList` class
- Add `FindCommand` command to handle the command
- Modify `Parser` to handle find command from user input

apollo-tan committed:
Merge branch 'branch-A-JavaDoc'

apollo-tan committed:
Merge branch 'branch-A-CodingStandard'

apollo-tan committed:
Merge branch 'branch-Level-9'

apollo-tan committed:
Replace wildcard import with explicit imports in Parser class

apollo-tan committed:
Implement barebones of GUI

apollo-tan committed:
Refactor code to use KukiShinobu instead of Duke class

apollo-tan committed:
Update logic of backend code to support new GUI
- Update execute method of Command class to return a String representing the response
	- Update all subsequent subclasses to support new String return in execute method
- Update KukiShinobu's getResponse method to obtain appropriate responses depending on the input

apollo-tan committed:
Update GUI to display response from all commands

- Update all commands to return a String representing the response from KukiShinobu
- Update methods in TaskList and Task to return a response String to actions taken rather than printing them directly to the console

apollo-tan committed:
Make some UI fixes and enhancements
- Update MainWindow to show welcome message when chat is first opened
- Update avatars to fit the actual characters that the chatbot represents

apollo-tan committed:
Merge branch 'branch-Level-10'

apollo-tan committed:
Main class: add assertion to ensure kukiShinobu is not null

The `Main` class sets the `kukiShinobu` instance in the `MainWindow` controller. However, there is no guarantee that `kukiShinobu` is properly initialized before the controller uses it.

To prevent potential `NullPointerException` errors and ensure the application runs smoothly, add an assertion to verify that `kukiShinobu` is not null.

Add an assertion statement after setting the `kukiShinobu` field in the controller to check that it is not null. This assertion will throw an error with a descriptive message if the field is null, aiding in debugging and preventing runtime issues.

Assertions are enabled with the `-ea` flag, which ensures that this check is only performed during development and testing.

No other changes are made; the focus is solely on improving reliability and debugging support.

apollo-tan committed:
Refactor `load` and `write` Methods: Extract File Handling and Task Parsing

The `load` and `write` methods handle multiple responsibilities, including file operations and task processing. This complexity can lead to maintenance challenges and reduced readability.

Extract the file reading and writing functionalities from the `load` and `write` methods into dedicated methods. Specifically, `readFileContent` handles reading file content, and `convertTasksToString` and `writeStringToFile` handle task serialization and file writing, respectively. Also, introduce `parseTask`, `parseTodo`, `parseDeadline`, and `parseEvent` methods to isolate task parsing logic.

Let's separate file handling and task parsing into individual methods to enhance readability and maintainability. By isolating concerns, each method becomes more focused and easier to test or modify independently.

Using helper methods is beneficial for managing complex operations and ensures that each method has a single responsibility.

No additional dependencies or structural changes are introduced.

apollo-tan committed:
**Remove Deprecated `Ui` Class and Associated Code**

- **Removed `Ui` class**: The `Ui` class, which provided terminal-based user interface functionalities, has been removed. This class was responsible for handling user input, displaying messages, and showing visual dividers in the console.

- **Updated `KukiShinobu` class**: The `KukiShinobu` class has been refactored to remove dependencies on the `Ui` class. The methods that previously interacted with `Ui` have been adjusted to work directly with the updated command execution and storage handling mechanisms.

- **Refactored `Command` class and subclasses**:
  - Modified the `Command` class and its subclasses to no longer require a `Ui` object.
  - Updated methods in `Command` and its subclasses to interact with `TaskList` and `Storage` directly.

- **Adjusted `getResponse` method**: The `getResponse` method in the `KukiShinobu` class has been updated to reflect the removal of `Ui` handling.

- **Removed `Ui` related imports**: All imports related to the `Ui` class have been removed from `KukiShinobu` and command classes.

- **Cleanup**: Removed unused `show` and `display` methods, and any associated functionality related to the terminal-based UI.

This refactor aims to streamline the application by removing deprecated UI components and simplifying the command handling process. The application now relies on updated methods for user interaction and storage management.

apollo-tan committed:
Refactor TaskList: Update `toString` and `findMatchingTasks` methods to use streams

The `toString` method and `findMatchingTasks` method in `TaskList` class currently use traditional loops and manual list manipulations.

Refactor these methods to leverage Java Streams for more concise and functional-style code.

- **Update `toString` method**: Replace manual loop with `IntStream` and `Collectors.joining` to generate a formatted string representation of tasks. This approach simplifies the code and improves readability.

- **Update `findMatchingTasks` method**: Replace manual list filtering with a stream-based approach, using `filter` and `collect` to create a list of tasks that match the keyword. This makes the method more expressive and aligns with modern Java practices.

This refactoring enhances code clarity and maintains consistency with functional programming principles.

apollo-tan committed:
Merge pull request #2 from apollo-tan/branch-A-Assertions

Main class: add assertion to ensure kukiShinobu is not null

apollo-tan committed:
Merge branch 'master' into branch-A-CodeQuality

apollo-tan committed:
Merge branch 'master' into branch-A-Streams

apollo-tan committed:
Merge pull request #3 from apollo-tan/branch-A-CodeQuality

Refactor Code Quality: Extract File Handling, Task Parsing, and Remove Deprecated UI

apollo-tan committed:
Merge branch 'master' into branch-A-Streams

apollo-tan committed:
Merge pull request #4 from apollo-tan/branch-A-Streams

Refactor TaskList: Update `toString` and `findMatchingTasks` Methods to Use Streams

apollo-tan committed:
Add support for tagging in Task and its subclasses

The Task class does not support tagging, which limits its ability to categorize and filter tasks.

Let's integrate a Tags class into Task to manage task tags effectively. Update constructors to accept and initialize tags. Adjust the Todo, Deadline and Event classes to utilise these constructors and include tag information in database storage.

Incorporate tags into the database string format to ensure comprehensive task records.

This approach enhances task management by ensuring tags are correctly handled and stored, improving task organization and retrieval.

apollo-tan committed:
Merge pull request #5 from apollo-tan/branch-C-Tagging-Barebones

Add support for tagging in Task and its subclasses

apollo-tan committed:
Enhance command error handling and documentation

Add Javadocs to improve clarity across multiple command classes.

Update constructors in AddDeadlineCommand, AddEventCommand, DeleteTaskCommand, MarkTaskCommand, and UnmarkTaskCommand to handle erroneous user input. Throw KukiShinobuException for invalid formats or missing data.

Move error handling for invalid user input from MainWindow class to Parser class. This centralizes error management and simplifies command execution in the UI layer.

Refactor to ensure consistent error handling and better separation of concerns. Using the Parser class for error handling streamlines the command parsing process and improves maintainability.

apollo-tan committed:
Merge pull request #6 from apollo-tan/A-MoreErrorHandling

Enhance command error handling and documentation

apollo-tan committed:
Add title and icon for greater personalisation

apollo-tan committed:
Merge pull request #7 from apollo-tan/branch-Final-Touches

Add title and icon for greater personalisation

apollo-tan committed:
Fix issue with Storage parsing emtpy text file

Modify parser of Storage to ignore empty text file rather than processing the empty string as a task, subsequently throwing an error since an empty string is not a valid format for a task

apollo-tan committed:
Update docs/README and add Ui.png

apollo-tan committed:
Merge pull request #8 from apollo-tan/branch-A-UserGuide

Add user guide a fix a bug

apollo-tan committed:
Remove errorneous line in docs/README.md

apollo-tan committed:
Merge pull request #9 from apollo-tan/branch-Fix-Readme

Remove errorneous line in docs/README.md

apollo-tan committed:
Update README.md

apollo-tan committed:
Merge pull request #10 from apollo-tan/branch-Update-Readme-1

Update README.md

apollo-tan committed:
Fix bug pertaining to tags

- Solve issue where tagless tasks are stored and parsed erroneously
  leading to loss of data or wrongly rendering the dates as tags
- Solve issue where listing todos doesn't display tags
- Update java docs of several methods to align with their current
  behaviour

apollo-tan committed:
Fix issue relating to index out of bounds

- Fix issue where specifying an index that is out of range for mark,
  unmark and delete commands lead to the program not responding
  appropraitely
- Create a new custom exception IndexOutOfBoundsKukiShinobuException to
  facilitate the above bugfix

apollo-tan committed:
Update docs/README to reflect new release

apollo-tan committed:
Update docs/README.md to point to latest release

apollo-tan committed:
Merge pull request #11 from apollo-tan/branch-Final-Bug-Fixes

Branch final bug fixes

apollo-tan committed:
Revamp README.md to reflect current state of code

apollo-tan committed:
Merge pull request #12 from apollo-tan/branch-Better-Readme

Revamp README.md to reflect current state of code

apollo-tan committed:
Fix calling list with empty tasklist  returns nothing

apollo-tan committed:
Update error message to be more helpful for event command

apollo-tan committed:
Update docs/README to reflect quirks of event command

apollo-tan committed:
Merge pull request #13 from apollo-tan/branch-Fix-Bugs-2

Fix some glaring bugs with regards to the bot

apollo-tan committed:
Add extensive test for important classes

- Add extensive tests for TaskList and Parser especially after modifications to the two classes
- Update README.md to reflect use of AI in generating test cases

apollo-tan committed:
Refactor code including removing unused methods

apollo-tan committed:
Revamp error handling of invalid commands

apollo-tan committed:
Merge pull request #14 from apollo-tan/branch-Better-Testing

Add more robust testing and refine code
